sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala

```diff
@@ -319,15 +319,7 @@ class SymmetricHashJoinStateManager(
       keyWithIndexToValueMetrics.numKeys,       // represent each buffered row only once
       keyToNumValuesMetrics.memoryUsedBytes + keyWithIndexToValueMetrics.memoryUsedBytes,
       keyWithIndexToValueMetrics.customMetrics.map {
-        case (s @ StateStoreCustomSumMetric(_, desc), value) =>
```

Comment:
didn't quite get why it worths the effort, you have add code either here or in the case class. no difference to me. 

or maybe you can save the effort by provide some default implementation.

Reply:
The main point is to avoid making changes here (in `SymmetricHashJoinManager`) whenever you add a new case class. You just add all the code required in the case class when a new one is added. It is easy to miss adding the code here and with this patch we can avoid exposing the subclasses of `StateStoreCustomMetric` here.

Reply:
+1
this is easier to maintain. exhaustive enumeration of subclasses is brittle to addition of subclasses.

Reply:
it would be even better if with some scala magic the default implementation could be provided. but its not obvious to me how, because this is calling the copy constructor which is not available in any trait.

Reply:
If we are also considering extension point from outside (like 3rd party state store provider) to add a new implementation of `StateStoreCustomMetric`, we should definitely not suppose the new implementation to be a case class having `desc` as field. I think this is the right way to do.